### ISSUE TITLE
feat: data-driven confidence calibration (isotonic regression)

### ISSUE THREAD

**Issue by kamilpajak** [kamilpajak/heisenberg#50]
## Context

Community [comment on #49](https://github.com/kamilpajak/heisenberg/issues/49#issuecomment-2777289518) suggests replacing our rule-based confidence caps with data-driven calibration:

1. **Isotonic regression** on `(LLM_confidence, actual_outcome)` pairs
2. **Bayesian blast-radius prior** with `P(infra | blast_radius > 0.25) ≈ 0.85`
3. **Reliability diagrams** for validation

These are sound ML techniques — but premature for our current stage.

## Why not now

Consulted zen (Gemini expert review) and Perplexity Deep Research. Both independently conclude:

**Isotonic regression requires hundreds of samples to be reliable.** With our current ~30 eval cases, it would severely overfit — fitting noise rather than true calibration curves. No published research favors data-driven calibration for n<100 (Perplexity, citing Google Rules of ML, arxiv:2509.15848).

**Bayesian priors hide a guess behind math.** Computing `P(infra | blast_radius)` requires historical data we don't have. Guessing the prior makes it a heuristic with unnecessary floating-point complexity. Our hard cap (`blast_radius > 0.5 + code blame → cap 39`) is simpler, explicit, and fully unit-testable (zen).

**Rule-based → data-driven is the standard production ML path.** Google's Rules of ML explicitly advises: "use simple heuristics to launch, then switch to ML once data is available" (Perplexity, citing developers.google.com/machine-learning/guides/rules-of-ml).

## Current state (V1 — rule-based)

Implemented in PR (pending merge on `feature/confidence-calibration` branch):

- 5 deterministic hard cap rules (blast radius, diff intersection, hidden infra evidence, etc.)
- Caps aligned with confidence rubric: 39 = uncertain, 49 = ambiguous
- 25 unit tests
- Live-validated on real Azure DevOps builds

### Prerequisites
- [ ] Grow eval corpus to 500+ labeled cases with `(predicted_confidence, actual_resolution)` pairs
- [ ] Log calibration telemetry: for every analysis, record `original_confidence`, `calibrated_confidence`, `calibration_reason`, and eventual manual resolution (code/infra/external)
- [ ] Build ground truth validation pipeline (from eval framework + Azure/GitHub live testing)

### Implementation
- [ ] Fit isotonic regression or Platt scaling on collected data
- [ ] Run in **shadow mode** alongside hard caps — compare Brier scores / log-loss
- [ ] Plot reliability diagrams (predicted confidence vs actual accuracy in decile bins)
- [ ] Replace hard caps only when data-driven approach demonstrably outperforms heuristics
- [ ] Consider per-failure-type calibration (separate models for timeout vs assertion vs network)

### Validation
- [ ] Reliability diagram should lie on the diagonal (well-calibrated)
- [ ] Brier score of V2 < Brier score of V1
- [ ] False positive rate of V2 ≤ false positive rate of V1

## References

- Google Rules of ML: "Use simple heuristics to launch" — https://developers.google.com/machine-learning/guides/rules-of-ml
- Guo et al. (2017): "On Calibration of Modern Neural Networks" — temperature scaling baseline
- Isotonic regression for calibration — sklearn.isotonic.IsotonicRegression
- arxiv:2509.15848 — rule-based vs data-driven comparison for small datasets
- Community comment: #49 (comment)

**Comment by Whatsonyourmind** [comment]
Two suggestions for the telemetry schema and evaluation plan:

**1. Log the full calibration context, not just the pair.**
Beyond `(predicted_confidence, actual_resolution)`, also log the features that fed into your hard-cap rules: `blast_radius`, `diff_intersection_score`, `n_failed / n_total`, `error_signature_hash`. When you later fit isotonic regression, you will want to analyze *which* feature regime the model struggles in. This also enables per-failure-type calibration without re-running old analyses.

**2. Use Brier skill score (BSS) rather than raw Brier score for V1 vs V2 comparison.**
Raw Brier score depends on the base rate of code-caused failures in your dataset. If 80% of failures are code-caused, a model that always predicts 80% gets a decent Brier score by doing nothing useful. BSS normalizes against this climatological baseline:

```
BSS = 1 - (Brier_model / Brier_climatology)
```

where `Brier_climatology = base_rate * (1 - base_rate)`. BSS > 0 means the model is better than always predicting the base rate. This gives a cleaner comparison between V1 and V2 because it removes the influence of class imbalance in your eval corpus.

For reliability diagrams: 10 bins (deciles) is standard but with 500 cases that is only 50 per bin. If your confidence distribution is skewed (many predictions near 0 or 100), some bins will be sparse. Adaptive binning (equal number of samples per bin rather than equal width) gives more stable calibration curves.

**Comment by Whatsonyourmind** [comment]
Hey @kamilpajak — how's the prediction data coming in? Once you have ~100 resolved predictions, the Brier skill score will start showing meaningful calibration patterns.

I have a calibration scoring API (Brier + log score + reliability diagrams) that can process your prediction history automatically — returns decomposed scores so you can see if the issue is discrimination vs reliability. Want me to send you a free API key to try it?
